### ISSUE TITLE
Unify API working with time duration

### ISSUE THREAD

**Issue by mihails-strasuns-sociomantic** [sociomantic-tsunami/ocean#272]
Right now there is a confusing mismatch between different ocean module working with timer events. For example, `ocean.io.select.client.TimerEvent` works with "seconds + milliseconds" pair but `ocean.io.select.client.TimerSet` works with microseconds. Under the hood it all gets converted to C binding pair of "seconds + nanoseconds".

I would suggest to either use `ocean.time.Time.TimeSpan` or introduce a simpler replacement for duration concept and introduce overloads for all APIs working with durations accepting that.

**Comment by david-eckardt-sociomantic** [comment]
Good idea. I suggest `ulong ms` for all timer values. Or is there a use case for a sub-ms timer?
This would also avoid the unnecessary complication of the interface that it is often not clear whether the separation in seconds and milliseconds requires the the ms value to be less than 1s. For example, is a time value of 3s + 1500ms (that is, 4.5s) valid, or does a function require 4s + 500ms instead?
Btw. in many cases the D functions reflect the inconsistency of the underlying C library functions:
- The `timeval` struct (`gettimeofday()`) uses µs,
- the `timespec` struct (`clock_gettime()`, timer event) uses ns, which ocean reduces to ms,
- `epoll_wait()` uses ms.

**Comment by mihails-strasuns-sociomantic** [comment]
Problem with `ulong ms` is that we won't be able to introduce non-breaking overloads in the API. Which is why I have been thinking about struct wrapper akin to `ocean.time.Time`

**Comment by gavin-norman-sociomantic** [comment]
I wonder if `timerfd_create` _really_ supports nanosecond resolution!?

But yeah, the idea of a time struct sounds sensible.

**Comment by david-eckardt-sociomantic** [comment]
Looking at the big time type and struct mess in C and D libraries I strongly prefer a primitive unsigned integer type and ms as the standard unit. I don't think adding yet another struct would be a good idea, it would just increase the convolution of time struct types.

> I wonder if `timerfd_create` really supports nanosecond resolution!?

The underlying timer probably does, but the epoll overhad defeats it.

**Comment by mihails-strasuns-sociomantic** [comment]
> I strongly prefer a primitive unsigned integer type

Well, the problem is that we simply can't have it without breaking half of ocean and downstream apps in an extremely hard to debug way. Also I'd much rather have `wait(seconds(1))` than `wait(1_000)`.

> I don't think adding yet another struct would be a good idea, it would just increase the convolution of time struct types.

Yeah multiplication of types would be messy. But looking at time module now, I think https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/time/Time.d#L62 is pretty much what is needed already. We can separate it into own module and deprecate all the unnecessary bloat in public fields and it should be enough to get things done.

**Comment by david-eckardt-sociomantic** [comment]
> Also I'd much rather have `wait(seconds(1))` than `wait(1_000)`.

Well, I really prefer `wait(1_000)`. I'm allergic to ticks.

**Comment by mihails-strasuns-sociomantic** [comment]
`1_000` of what? :trollface:

**Comment by david-eckardt-sociomantic** [comment]
Milliseconds, the natural time unit of reason.

**Comment by gavin-norman-sociomantic** [comment]
>1_000 of what? :trollface:

1,000 ticks, obviously! :trollface: :trollface:

**Comment by mihails-strasuns-sociomantic** [comment]
> 1,000 ticks, obviously!

I am reasonably sure it was meant to be `double seconds`.

**Comment by david-eckardt-sociomantic** [comment]
I could make piece with a single-member struct carrying the time in µs (rather than obscure ticks) and conversion methods.
A point is that, while this is convenient to express constant time values it makes calculating difficult. At least calculating a time difference and summing times up are common things to do. With plain integers this is straightforward, with structs it is not, and operator overloading can easily cause [code bloat](https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/time/Time.d#L128-L275).

**Comment by mihails-strasuns-sociomantic** [comment]
> I could make piece with a single-member struct carrying the time in µs

That sounds reasonable (assuming µs is max precision we ever need).

I would even suggest to make µs integer field public so that one can do arbitrary calculation without operator overloading or any other mess. Basically the struct would only exist so that you can overload on it and know for sure the number is indeed amount of µs.

**Comment by gavin-norman-sociomantic** [comment]
>I would even suggest to make µs integer field public so that one can do arbitrary calculation without operator overloading or any other mess. Basically the struct would only exist so that you can overload on it and know for sure the number is indeed amount of µs.

Yep, this is what I had in mind too. The setter methods will just be for convenience.

**Comment by leandro-lucarella-sociomantic** [comment]
I know some will hate me, but I suggest using seconds as a float everywhere (and do the pertinent conversions when needed). It could be a struct too with convenient conversion functions to the popular C API structs, but using floats as a default type is quite clear, simple and flexible (and times usually are not really that precise to need an integer type, rounding shouldn't be an issue).

**Comment by mihails-strasuns-sociomantic** [comment]
How about no? :)

1) Slower than integers on all arithmetic operations
2) Imprecise (trying to arm timer on microsecond precision should be enough to trigger it)
3) Requires conversions each time you interact with C libraries
4) Doesn't help with convenience because you would still want to use `wait(ms(100))` over `wait(0.0000001)`

**Comment by david-eckardt-sociomantic** [comment]
> I know some will hate me

:rage:

Everything @mihails-strasuns-sociomantic said plus: Time values need absolute exact precision that is straightforward to understand and work with.
- There is no decimal one- or two-digit milliseconds value that can be represented exactly as a binary floating point number. Adding number representation fuzziness by in- or decrementing an exact time value by 1ms or 10ms is IMHO unacceptable.
- Even though one can reasonably argue, in reality there is no difference between 86400s (one day) and 86400.000001s -- the physical clock has a relative error just like floating-point numbers -- it is confusing and exceeds most people's logical understanding that `(t + dt) - t ≠ dt`.

**Comment by david-eckardt-sociomantic** [comment]
> There is no decimal one- or two-digit milliseconds value that can be represented exactly as a binary floating point number.

(I bet a free lunch.)

**Comment by leandro-lucarella-sociomantic** [comment]
> Slower than integers on all arithmetic operations

Really? So timer arithmetic will slow down your programs... :roll_eyes: (I think this is completely a non-issue).

> Imprecise (trying to arm timer on microsecond precision should be enough to trigger it)

False. It will be imprecise only if you need to set up a timer with more than 7 significant digits, which sounds crazy. Do you really see someone doing `timer(0.0012345678)`? You probably always need just 1, 2, 3 at tops significant digits.

> Requires conversions each time you interact with C libraries

Again, a non issue performance wise, and as long as we go with an unified format for specifying times for timers, we need to convert because the whole point of this issue is there are several ways to specify timers in C APIs.

> Doesn't help with convenience because you would still want to use wait(ms(100)) over wait(0.0000001)

If you want to use more symbolic names, yes, but I would actually do `wait(1e-6)` to specify `us` and `100e-3` to specify `ms` (or just `0.1` in this case actually). Don't get me wrong, I'm all for adding more symbolic names, they are totally worth it, but for me `float` in seconds (or `double` or `real` if you are that concerned about precision) is a perfect match for this, as one usually always work close to seconds in most cases (100ms -> 0.1s, tcp keepalive are in seconds for example, etc.).

> There is no decimal one- or two-digit milliseconds value that can be represented exactly as a binary floating point number. Adding number representation fuzziness by in- or decrementing an exact time value by 1ms or 10ms is IMHO unacceptable.

Again, this is not true. I mean, technically, or mathematically, it is, but it is not true in the case of timers, where talking about a precision of more than a couple of digits is crazy and flawed in the first place. Every value up to ~7 significant digits, no matter how big or small it is as long as it is between ~1e-127 and ~1e127 can be perfectly represented in practical terms!

Again, if you need more precision that a couple of digits, then you'll be deceived by the OS anyway, because the timers are never precise, so the potential lack of precision of a float in terms of significant digits is just a good way to make it obvious that no precise value will be respected by the timers.

This is probably my last argument, because is not that important and is better not to waste time on a flame war :)

PS: Just as a practical proof:
```d
import std.stdio;
void main()
{
    float ms = 0.00123456;
    float us = 0.00000123456;
    writefln("ms=%s as int in nanoseconds=%d", ms, cast(int)(ms*1e9));
    writefln("us=%s as int in picoseconds=%d", us, cast(int)(us*1e12));
}
```
Look ma'! It works!
```
ms=0.00123456 as int in nanoseconds=1234560
us=1.23456e-06 as int in picoseconds=1234560
```

**Comment by leandro-lucarella-sociomantic** [comment]
One last anecdotal thing, I didn't just had a revelation, I took this idea from well established projects. This is already used by Python (for example: https://docs.python.org/3.6/library/select.html#select.select) and even `libev`, *a high performance full-featured event loop written in C* (http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#code_ev_timer_code_relative_and_opti / http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#TIME_REPRESENTATION)

**Comment by david-eckardt-sociomantic** [comment]
> Again, this is not true. I mean, technically, or mathematically, it is, but it is not true in the case of timers, where talking about a precision of more than a couple of digits is crazy and flawed in the first place. Every value up to ~7 significant digits, no matter how big or small it is as long as it is between ~1e-127 and ~1e127 can be perfectly represented in practical terms!

That is all correct, but it is not what I mean. My point is: Integer numbers reflect the behaviour of the underlying mechanism, which is that there is a clock ticking in discrete steps. They convey a correct understanding of what happens and keep the promises they make -- precise values with an absolute precision. You can always add or subtract 1µs or 27ms or one week and 3.14159s or whatever, your logical time value will exactly be what you expect it to. There are no surprises -- I haven't met a person yet who was surprised by the fact that a timer isn't precise to the microsecond over a week.
Floating-point, on the other hand is handy _in many practical situations_ but counter-intuitive, breaks promises of numbers in computing and therefore do in practice cause confusion. `(t + dt) - t ≠ dt` in a program calculation breaks the promise of understandable math. `1.E-30` is possible so it is unclear what the absolute precision is, which is important to know. 

We should use integers because they reflect what happens and intuitively lead to the correct expectations. Floating-point convenience causes bad surprises and abstraction inversion, sooner or later.

**Comment by mihails-strasuns-sociomantic** [comment]
Please for the sake of all good don't bring Python into this argument :) I will give it some respect for distinguishing integers and floats at all, but scripting languages have completely different criteria of performance and usability.

Libev would be more relevant, but there is also one widely known C library WE ACTUALLY USE that doesn't do that.

And this:

> but I would actually do wait(1e-6) to specify us and 100e-3 to specify ms

.. makes me adamant on not letting your proposal through just for the sake of not ever seeing anyone write  some crazy code like that ;)

**Comment by david-eckardt-sociomantic** [comment]
Oh, I would be glad to `wait(150E-3)` ;)

**Comment by mathias-lang-sociomantic** [comment]
@leandro-lucarella-sociomantic : I think using float for any time of time-keeping is a terrible idea. There are enough [horror stories](http://www-users.math.umn.edu/~arnold/disasters/patriot.html) out there to convince anyone of this.

It's not only about using it as a single, independent value for representation, but the operations you'll perform on them. Just try to compute the total run-time used by a callback called every second, over the course of a week. How long before you need more than those 7 digits of "precision" ?
